File: scripts/html_utils.py
"""
HTML prettification utilities for the Sparge editor view.
"""
import re


def prettify_html(raw: str) -> str:
    """Prettify HTML for the editor while preserving inline element adjacency.

    BeautifulSoup prettify() puts every element on its own line, including
    inline elements like <b>, <strong>, <em>. This hides whether </b> was
    immediately adjacent to the following character in the original HTML — a
    crucial signal for understanding the md_notation_in_text warning.

    Example: <b>Bob Kowalski</b>(Imperial College London) would display as:
        <b>
         Bob Kowalski
        </b>
        (Imperial College London)
    making it impossible to see that </b> and ( are adjacent.

    After this function:
        <b>Bob Kowalski</b>(Imperial College London)
    The adjacency is visible; the user can immediately understand why
    html2text produces **Bob Kowalski**(Imperial with no space.

    When there IS a real space — <b>text</b> (more) — the closing tag and
    the following content stay on separate lines, preserving the distinction.

    Also uses html.parser (not lxml) to avoid double-encoding non-ASCII
    characters (em dashes, curly quotes) via lxml's charset sniffing.
    """
    from bs4 import BeautifulSoup, NavigableString, Tag

    soup = BeautifulSoup(raw, 'html.parser')

    # WORD JOINER (U+2060) — invisible, won't appear in normal HTML text.
    # We prefix the following sibling with it to mark adjacent inline elements.
    MARKER = '\u2060'

    INLINE_TAGS = frozenset({
        'b', 'strong', 'em', 'i', 'code', 'a', 'abbr',
        'cite', 'q', 's', 'u', 'del', 'ins', 'mark', 'small', 'sub', 'sup',
    })

    # Mark inline elements whose closing tag is immediately adjacent to a
    # non-whitespace character in the original HTML. We prefix the following
    # NavigableString sibling with MARKER so prettify() preserves the info
    # (as a leading character on its own line) and we can rejoin them after.
    for tag in list(soup.find_all(list(INLINE_TAGS))):
        if not isinstance(tag, Tag):
            continue
        if tag.find_parent(['pre', 'code']):
            continue
        sib = tag.next_sibling
        if isinstance(sib, NavigableString) and sib and not sib[0].isspace():
            sib.replace_with(NavigableString(MARKER + str(sib)))

    content = soup.prettify()

    # Garbling detection — ÃÂÃÂ is the signature of lxml double-encoding
    # UTF-8 as Latin-1. Should not happen with html.parser, but guard anyway.
    if 'ÃÂÃÂ' in content or '\xc3\x82' in content:
        return raw

    # Inline element pattern — matches tag names that should stay on one line.
    _INLINE = r'(?:b|strong|em|i|code|a|abbr|cite|q|s|u|del|ins|mark|small|sub|sup)'

    # Step 1: Collapse simple text inside inline elements to a single line.
    #   Before: <b>↵   Bob Kowalski ↵  </b>
    #   After:  <b>Bob Kowalski</b>
    # Only collapses when the content is plain text (no child tags).
    content = re.sub(
        rf'(<(?:{_INLINE})(?:\s[^>]*)?>)\n[ \t]*(.*?)\n[ \t]*(</(?:{_INLINE})>)',
        lambda m: m.group(1) + m.group(2) + m.group(3),
        content,
        flags=re.IGNORECASE,
    )

    # Step 2: Rejoin closing tags with MARKER-prefixed content (adjacent in original).
    #   Before: </b>↵   ⁠(Imperial College London)
    #   After:  </b>(Imperial College London)
    # The MARKER identifies lines where the original had no space before content.
    content = re.sub(
        rf'(</(?:{_INLINE})>)\n[ \t]*{re.escape(MARKER)}',
        r'\1',
        content,
        flags=re.IGNORECASE,
    )

    # Clean up any remaining MARKER characters (e.g. inside nested inline elements
    # where Step 1 did not collapse the outer element's content).
    content = content.replace(MARKER, '')

    return content

File: server/src/main/java/io/sparge/server/HtmlUtils.java
package io.sparge.server;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.Node;
import org.jsoup.nodes.TextNode;

import java.nio.charset.StandardCharsets;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * HTML prettification utilities for the Sparge editor view.
 * Mirrors scripts/html_utils.py: prettify_html().
 *
 * Uses the same MARKER trick as Python: U+2060 WORD JOINER is prepended to
 * sibling text immediately adjacent to inline closing tags, then Jsoup
 * pretty-prints (with UTF-8 charset so U+2060 passes through unescaped),
 * then regex rejoins the marked adjacencies.
 */
public final class HtmlUtils {

    /** U+2060 WORD JOINER — invisible, not HTML-special, survives Jsoup UTF-8 output. */
    private static final String MARKER = "\u2060";

    private static final Set<String> INLINE_TAGS = Set.of(
            "b", "strong", "em", "i", "code", "a", "abbr",
            "cite", "q", "s", "u", "del", "ins", "mark", "small", "sub", "sup"
    );

    private static final String INLINE_RE =
            "(?:b|strong|em|i|code|a|abbr|cite|q|s|u|del|ins|mark|small|sub|sup)";

    // Step 1: collapse <b>\n  TEXT\n </b> to <b>TEXT</b>
    private static final Pattern COLLAPSE = Pattern.compile(
            "(?i)(<(?:" + INLINE_RE + ")(?:\\s[^>]*)?>)\\n[ \\t]*([^\\n]*)\\n[ \\t]*(</(?:" + INLINE_RE + ")>)"
    );

    // Step 2: join </b>\n  ⁠text → </b>text (MARKER at start of following line)
    private static final Pattern JOIN_ADJACENT = Pattern.compile(
            "(?i)(</(?:" + INLINE_RE + ")>)\\n[ \\t]*" + Pattern.quote(MARKER)
    );

    private HtmlUtils() {}

    /**
     * Prettify HTML for the editor while preserving inline element adjacency.
     * Mirrors scripts/html_utils.py: prettify_html().
     */
    public static String prettifyHtml(String raw) {
        if (raw == null || raw.isEmpty()) return raw;

        Document doc = Jsoup.parse(raw);
        // UTF-8 output: non-ASCII chars (including U+2060) pass through unescaped
        doc.outputSettings()
                .charset(StandardCharsets.UTF_8)
                .prettyPrint(true)
                .indentAmount(1);

        // Mark inline elements whose closing tag is immediately adjacent to non-whitespace
        for (String tagName : INLINE_TAGS) {
            for (Element el : doc.select(tagName)) {
                if (el.parents().stream()
                        .anyMatch(p -> p.tagName().equals("pre") || p.tagName().equals("code"))) {
                    continue;
                }
                Node sibling = el.nextSibling();
                if (sibling instanceof TextNode tn) {
                    String text = tn.getWholeText();
                    if (!text.isEmpty() && !Character.isWhitespace(text.charAt(0))) {
                        tn.text(MARKER + text.stripLeading());
                    }
                }
            }
        }

        String content = doc.outerHtml();

        // Garbling detection
        if (content.contains("ÃÂÃÂ") || content.contains("\u00c3\u0082")) {
            return raw;
        }

        // Step 1: collapse inline element text to single line
        content = COLLAPSE.matcher(content).replaceAll("$1$2$3");

        // Step 2: rejoin adjacent closing tags with MARKER-prefixed content
        content = JOIN_ADJACENT.matcher(content).replaceAll("$1");

        // Clean up remaining MARKER characters
        content = content.replace(MARKER, "");

        return content;
    }
}
